src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+{
+    /// <summary>
+    /// A single gate for code that is adding work to <see cref="_projectsToLoadAndReload" />. This is just we don't have code simultaneously trying to load and unload solutions at once.
+    /// </summary>
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly ILspServices _lspServices;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private readonly IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
+    private readonly ProjectTargetFrameworkManager _projectTargetFrameworkManager;
+    private readonly ILoggerFactory _loggerFactory;
+
+    public bool SupportsMutation => true;
+
+    // ILspLogger rationalize
+    ILogger iLogger;
+
+    public LspFileBasedProgramWorkspace(
+        ILspServices lspServices,
+        IFileChangeWatcher fileChangeWatcher,
+        [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+        ProjectTargetFrameworkManager projectTargetFrameworkManager,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        HostServices hostServices) : base(hostServices, WorkspaceKind.FileBasedPrograms)
+    {
+        _lspServices = lspServices;
+        _fileChangeWatcher = fileChangeWatcher;
+        _dynamicFileInfoProviders = dynamicFileInfoProviders;
+        _projectTargetFrameworkManager = projectTargetFrameworkManager;
+        _loggerFactory = loggerFactory;
+        _projectsToLoadAndReload = new AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)>(
+            TimeSpan.FromMilliseconds(100),
+            LoadOrReloadProjectsAsync,
+            equalityComparer: null,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            CancellationToken.None);
+        iLogger = loggerFactory.CreateLogger<LspFileBasedProgramWorkspace>();
+    }
+
+    public async Task<Document?> AddMiscellaneousDocumentAsync(Uri uri, SourceText documentText, string languageId, ILspLogger logger, CancellationToken cancellationToken)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+        var documentId2 = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath);
+        var document2 = this.CurrentSolution.GetDocument(documentId2.FirstOrDefault());
+        if (document2 != null)
+        {
+            return document2;
+        }
```

Comment:
this needs doc'ing.  why are we returning an existing document if we were asked to add a doc.

Reply:
planning to delete this.